plugins/dell-k2/fu-dell-k2-plugin.c

```diff
@@ -0,0 +1,360 @@
+/*
+ * Copyright 2024 Dell Technologies
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR MIT
+ */
+
+#include "config.h"
+
+#include "fu-dell-k2-common.h"
+#include "fu-dell-k2-plugin.h"
+
+/* register the firmware types */
+#include "fu-dell-k2-dpmux-firmware.h"
+#include "fu-dell-k2-ilan-firmware.h"
+#include "fu-dell-k2-pd-firmware.h"
+#include "fu-dell-k2-rtshub-firmware.h"
+
+struct _FuDellK2Plugin {
+	FuPlugin parent_instance;
+};
+
+G_DEFINE_TYPE(FuDellK2Plugin, fu_dell_k2_plugin, FU_TYPE_PLUGIN)
+
+static gboolean
+fu_dell_k2_plugin_create_node(FuPlugin *plugin, FuDevice *device, GError **error)
+{
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	locker = fu_device_locker_new(device, error);
+	if (locker == NULL)
+		return FALSE;
+
+	fu_plugin_device_add(plugin, device);
+	return TRUE;
+}
+
+static gboolean
+fu_dell_k2_plugin_device_add(FuPlugin *plugin, FuDevice *device, GError **error)
+{
+	FuDellK2BaseType dock_type;
+	FuDevice *ec_device = fu_plugin_cache_lookup(plugin, "ec");
+	guint16 vid = fu_usb_device_get_vid(FU_USB_DEVICE(device));
+	guint16 pid = fu_usb_device_get_pid(FU_USB_DEVICE(device));
+
+	/* cache this device until dock type is seen */
+	if (ec_device == NULL) {
+		const gchar *key;
+		key = g_strdup_printf("USB\\VID_%04X&PID_%04X", vid, pid);
+		fu_plugin_cache_add(plugin, key, device);
+		return TRUE;
+	}
+
+	/* dock type according to ec */
+	dock_type = fu_dell_k2_ec_get_dock_type(ec_device);
+	if (dock_type == FU_DELL_K2_BASE_TYPE_UNKNOWN) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_READ,
+			    "can't read base dock type from EC");
+		return FALSE;
+	}
+
+	/* dell devices */
+	if (vid != DELL_VID) {
+		g_debug("ignoring unknown vid: %04X", vid);
+		return TRUE;
+	}
+
+	/* RTS usb hub devices */
+	switch (pid) {
+	case DELL_K2_USB_RTS5480_GEN1_PID:
+	case DELL_K2_USB_RTS5480_GEN2_PID:
+	case DELL_K2_USB_RTS5485_GEN2_PID:
+		g_autoptr(FuDellK2RtsHub) hub_device = NULL;
+		g_autoptr(FuDeviceLocker) locker = NULL;
+
+		hub_device = fu_dell_k2_rtshub_new(FU_USB_DEVICE(device), dock_type);
```

Comment:
just make this an `if` rather than a switch. The g_autoptr is getting very confused as it's not in a {} section.